DB/src/main/java/io/deephaven/db/tables/utils/ParquetTools.java

```diff
@@ -0,0 +1,409 @@
+        final File[] absoluteDestinations =
+                Arrays.stream(destinations)
+                        .map(File::getAbsoluteFile)
+                        .toArray(File[]::new);
+        final File[] firstCreatedDirs =
+                Arrays.stream(absoluteDestinations)
+                        .map(ParquetTools::prepareDestinationFileLocation)
+                        .toArray(File[]::new);
+
+        for (int i = 0; i < sources.length; i++) {
+            final Table source = sources[i];
+            try {
+                writeParquetTableImpl(source, tableDefinition, codecName, destinations[i], groupingColumns);
+            } catch (RuntimeException e) {
+                for (final File firstCreatedDir : firstCreatedDirs) {
+                    if (firstCreatedDir == null) {
+                        continue;
+                    }
+                    log.error("Error in table writing, cleaning up potentially incomplete table destination path starting from " +
+                            firstCreatedDir.getAbsolutePath(), e);
+                    FileUtils.deleteRecursivelyOnNFS(firstCreatedDir);
+                }
+                throw e;
+            }
+        }
+    }
+
+    /**
+     * Write out tables to disk.
+     *
+     * @param sources source tables
+     * @param tableDefinition table definition
+     * @param destinations destinations
+     */
+    public static void writeTables(@NotNull final Table[] sources,
+                                   @NotNull final TableDefinition tableDefinition,
+                                   @NotNull final File[] destinations) {
+        writeParquetTables(sources, tableDefinition, defaultPerquetCompressionCodec, destinations,
+                tableDefinition.getGroupingColumnNamesArray());
+    }
+
+    /**
+     * Deletes a table on disk.
+     *
+     * @param path path to delete
+     */
+    public static void deleteTable(File path) {
+        FileUtils.deleteRecursivelyOnNFS(path);
+    }
+
+    private static Table readTableFromSingleParquetFile(
+            @NotNull final File sourceFile,
+            @NotNull final ParquetInstructions readInstructions,
+            @NotNull final TableDefinition tableDefinition) {
+        final TableLocationProvider locationProvider = new ReadOnlyLocalTableLocationProviderByParquetFile(
+                StandaloneTableKey.getInstance(),
+                sourceFile,
+                false,
+                TableDataRefreshService.getSharedRefreshService(),
+                readInstructions);
+        return new SimpleSourceTable(tableDefinition.getWritable(), "Read single parquet file from " + sourceFile,
+                RegionedTableComponentFactoryImpl.INSTANCE, locationProvider, null);
+    }
+
+    private static Class<?> dbArrayType(final Class<?> componentTypeFromParquet) {
+        if (componentTypeFromParquet == null) {
+            return null;
+        }
+        if (componentTypeFromParquet.equals(int.class)) {
```

Comment:
I tried to arranged them by "expected frequency of call from high to low".
Shouldn't the code in SimpleTypeMap be generated?  Dispatching by map lookup sounds lame.  The optimizer can work chains of ifs like that (or switch statements) nicely, I don't think it can do much with a map.

Reply:
I'm still arguing for the map. Or at least a predictable order (size? alphabetic?) and the use of reference equality.